controlplane/kubeadm/internal/controllers/controller.go

```diff
@@ -329,6 +329,19 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
 	// source ref (reason@machine/name) so the problem can be easily tracked down to its source machine.
 	conditions.SetAggregate(controlPlane.KCP, controlplanev1.MachinesReadyCondition, ownedMachines.ConditionGetters(), conditions.AddSourceRef(), conditions.WithStepCounterIf(false))
 
+	// Ensure all required labels exist on the controlled Machines.
+	// This logic is needed to add the `cluster.x-k8s.io/control-plane-name` label to Machines
+	// which were created before the `cluster.x-k8s.io/control-plane-name` label was introduced.
+	// NOTE: Changes will be applied to the Machines in reconcileControlPlaneConditions.
+	// NOTE: cluster.x-k8s.io/control-plane is already set at this stage (it is used when reading controlPlane.Machines).
+	// TODO(sbueringer): Drop the following code with v1.4 after all existing Machines are guaranteed to have the new label.
```

Comment:
Note: Turns out we can't just drop it with v1.4 as we have no guarantees that everyone is upgrading v1.2 => v1.3 => v1.4

I would like to get rid of the logic at some point though. Just to keep the complexity of our already complex code as low as possible. Maybe we can drop it 1-2 years for now? (with some disclaimer in the corresponding .0 release notes)

Reply:
I think this isn't only for adding the label on upgrade but also for ensuring the label gets re-applied if someone removes it or changes its value.
So I will drop the todo

Reply:
Yup but I think this is only necessary for existing clusters? All new MachineSets will already have the labels.

Or do we support that folks are manually adding MachineSets below our MachineDeployments?

Ah I remember, this is about handling what happens if folks manually remove the labels?

I'm fine with keeping this, although the complexity that we have in the MD controller is really concerning already.

I wonder if the approach "what if a user deletes this label / ownerRef / field / ..." is something that we can sustain. There is just data that we have to be able to rely on otherwise all bets are off, e.g. what if a user removes the cluster label, ...

(to be clear, I'm fine with keeping this logic, just wondering how sustainable all this is longterm)